File: houseRobber3.cpp
#include <bits/stdc++.h>
using namespace std;


//Definition for treeNode
struct TreeNode{
   int val;
   TreeNode*left;
   TreeNode*right;
   TreeNode(int x): val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    pair<int, int> solve(TreeNode* root)
    {

        if(root==NULL)
        return {0,0};

         auto p1=solve(root->left);
        auto p2=solve(root->right);

        pair<int, int> p;
        p.first=root->val+p1.second+p2.second;
         p.second=max(p1.first, p1.second)+max(p2.first, p2.second);

         return p;
       
    }
    int rob(TreeNode* root) {
       pair<int,int>p = solve(root);

       return max(p.first,p.second);
    }
};

File: houseRobber3.py
# Definition for TreeNode
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        p = self.solve(root)
        return max(p[0], p[1])

    def solve(self, root: TreeNode) -> tuple:
        if root is None:
            return (0, 0)
        p1 = self.solve(root.left)
        p2 = self.solve(root.right)
        p = (0, 0)
        p = (root.val + p1[1] + p2[1], 
             max(p1[0], p1[1]) + max(p2[0], p2[1]))
        return p
